lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -720,8 +733,11 @@ bool firrtl::areTypesEquivalent(FIRRTLType destFType, FIRRTLType srcFType,
   if (!destType || !srcType)
     return destFType == srcFType;
 
-  // Type constness must match for equivalence
-  if (destType.isConst() && !srcType.isConst() && !srcOuterTypeIsConst)
+  bool srcIsConst = srcOuterTypeIsConst || srcFType.isConst();
+  bool destIsConst = destOuterTypeIsConst || destFType.isConst();
+
+  // A 'const' ground destination requires a 'const' source.
+  if (destType.isGround() && destIsConst && !srcIsConst)
     return false;
```

Comment:
Maybe we should keep all the ground-related logic grouped at the end?  The current logic down below sort of looks wrong to me if this were to fall through:
```c++
  // Ground types can be connected if they are the same, or the source type is
  // a const version of the destination type.
  return destType == srcType || destType == srcType.getConstType(false);
```
Should this be something like:
```c++
if (destIsConst && !srcIsConst)
    return false;
return destType.getConst(false) == srcType.getConst(false);

// OR:

auto realDestType = destType.getConstType(destIsConst);
return realDestType == srcType.getConstType(true) ||  realDestType == srcType.getConstType(false);
```

Reply:
I grouped the ground stuff at the end, and I also moved the reset-type tests below the aggregates as well.